### DISCUSSION TITLE
SHOP3 and dynamic planning

### DISCUSSION THREAD

**Discussion by pdenno** [shop-planner/shop3#145] [Q&A]
I have an application where I think dynamic planning might be a solution. About 10 years ago I used SHOP2, so that is where I began my search. I see that you (RPG) wrote an interesting paper on on-line POMDPs and that piqued my interest further. Poking around in the code however, I don't actually see mention of monte carlo tree search or hooks for using SHOP in a dynamic setting with uncertain data. Is that work in progress? Abandoned? Or something I need to develop for my particular application? 

I haven't had luck building the documentation. There seems to be only one texinfo in the distribution and the make process is looking for more.

**Comment by rpgoldman** [comment]
> I have an application where I think dynamic planning might be a solution. About 10 years ago I used SHOP2, so that is where I began my search. I see that you (RPG) wrote an interesting paper on on-line POMDPs and that piqued my interest further. Poking around in the code however, I don't actually see mention of monte carlo tree search or hooks for using SHOP in a dynamic setting with uncertain data. Is that work in progress? Abandoned? Or something I need to develop for my particular application?

I'm afraid that the MCTS code is still a work in progress. I have been working on using it with a board game, and haven't released it yet.

We have used SHOP in various dynamic contexts without explicit uncertainty by hooking it to executives of one sort or another and integrating replanning or plan repair, and the work on plan repair is available (although the plan repair badly needs a refactoring).
> 
> I haven't had luck building the documentation. There seems to be only one texinfo in the distribution and the make process is looking for more.

There may be a problem with the Makefile -- I don't know that it's really been tested by anyone but me.  If you have SBCL, the Makefile should load SHOP, then extract the docstrings into a number of individual Texinfo files, and incorporate them into the manual.  The library -- which I adapted from SB-TEXINFO -- only works on SBCL, I'm afraid (although SHOP works fine with SBCL, CCL and Allegro, at least).

If you can't get the manual to build, please file an issue about it, with a transcript, and I'll try to get it fixed up.

**Reply by pdenno** [reply]
Regarding the docs, I'm using sbcl, but haven't been coding in CL lately, so I could be wrong, but I had to clone the repository for sb-texinfo under the :tree defining my asdf source registry. With that done, it still had problems: It couldn't find an .asd file for plan-grapher.

FWIW, last week I wrote some Clojure to store the elements of SHOP domain definitions in a datalog DB. [here](https://github.com/pdenno/schedulingTBD/blob/main/src/server/scheduling_tbd/shop.clj)

**Reply by rpgoldman** [reply]
You shouldn't have had to clone any additional libraries if you did `git submodule init; git submodule update`.  That should give you all the required libraries in the `jenkins/ext/` subdirectory.

**Reply by rpgoldman** [reply]
Add that tree to your ASDF configuration.  Here's my (somewhat fussy) configuration for comparison:

```
(defparameter +shop3-source-registry+
  ;; I *hate* logical pathnames...
  (labels ((rewrite (x)
             (etypecase x
               (symbol x)
               (string (translate-logical-pathname x))
               (cons
                (if (eq (car x) :exclude)
                    x
                    (cons (rewrite (car x))
                          (rewrite (cdr x))))))))
    (rewrite
     `(:source-registry
       (:tree "site-lisp:shop;shop3;")
       (:tree 
        "site-lisp:shop;jenkins;ext;")
       :inherit-configuration)))
  "The s-expression used to initialize the source registry
when working on SHOP3.")
```

If you set up the logical pathname "site-lisp" something like that should work for you.
